packages/frontend-app-api/src/wiring/parameters.ts

```diff
@@ -0,0 +1,131 @@
+// as the Config interface makes it quite hard for us otherwise.
+/** @internal */
+export function readAppExtensionParameters(
+  rootConfig: Config,
+): Partial<ExtensionInstanceParameters>[] {
+  const arr = rootConfig.getOptional('app.extensions');
+  if (!Array.isArray(arr)) {
+    if (arr === undefined) {
+      return [];
+    }
+    // This will throw, and show which part of config had the wrong type
+    rootConfig.getConfigArray('app.extensions');
+    return [];
+  }
+
+  return arr.map((value, index) => {
+    function errorMsg(msg: string, key?: string, prop?: string) {
+      return `Invalid extension configuration at app.extensions[${index}]${
+        key ? `[${key}]` : ''
+      }${prop ? `.${prop}` : ''}, ${msg}`;
+    }
+
+    if (typeof value === 'string') {
+      return { id: value };
+    } else if (
+      typeof value !== 'object' ||
+      value === null ||
+      Array.isArray(value)
+    ) {
+      throw new Error(errorMsg('must be a string or an object'));
+    }
+
+    const keys = Object.keys(value);
+    if (keys.length !== 1) {
+      const joinedKeys = `"${keys.join('", "')}"`;
+      throw new Error(errorMsg(`must have exactly one key, got ${joinedKeys}`));
+    }
+
+    const key = keys[0];
+    const obj = value[key];
+    if (typeof obj === 'boolean') {
+      return { id: key, disabled: !obj };
+    }
+
+    if (typeof obj !== 'object' || obj === null || Array.isArray(obj)) {
+      throw new Error(errorMsg('must be an object', key));
+    }
+    const at = obj.at;
+    if (at !== undefined && typeof at !== 'string') {
+      throw new Error(errorMsg('must be a string', key, 'at'));
+    }
+    const disabled = obj.disabled;
+    if (disabled !== undefined && typeof disabled !== 'boolean') {
+      throw new Error(errorMsg('must be a boolean', key, 'disabled'));
+    }
+    const extension = obj.extension;
+    if (extension !== undefined && typeof extension !== 'string') {
+      throw new Error(errorMsg('must be a string', key, 'extension'));
+    }
+    if (extension) {
+      throw new Error('TODO: implement extension resolution');
+    }
+    return { id: key, at, disabled, config: obj.config /* validate later */ };
+  });
+}
+
+/** @internal */
+export function mergeExtensionParameters(
+  base: ExtensionInstanceParameters[],
+  overrides: Array<Partial<ExtensionInstanceParameters>>,
+): ExtensionInstanceParameters[] {
+  const extensionInstanceParams = base.slice();
+
+  for (const overrideParam of overrides) {
+    const existingParamIndex = extensionInstanceParams.findIndex(
+      e => e.id === overrideParam.id,
+    );
+    if (existingParamIndex !== -1) {
+      const existingParam = extensionInstanceParams[existingParamIndex];
+      if (overrideParam.at) {
+        existingParam.at = overrideParam.at;
+      }
+      if (overrideParam.extension) {
+        // TODO: do we want to reset config here? it might be completely
+        // unrelated to the previous one
```

Comment:
especially if the new extension has a strict schema that rejects unknown config keys - yeah i think we may have to do this (in a later PR)